src/ophyd_async/core/_detector.py

```diff
@@ -313,12 +331,17 @@ async def prepare(self, value: TriggerInfo) -> None:
     @AsyncStatus.wrap
     async def kickoff(self):
         assert self._trigger_info, "Prepare must be called before kickoff!"
-        if self._iterations_completed >= self._trigger_info.iteration:
+        if self._frames_collected >= self._total_frames:
             raise Exception(
                 f"Kickoff called more than the configured number of "
-                f"{self._trigger_info.iteration} iteration(s)!"
+                f"{self._total_frames} iteration(s)!"
             )
-        self._iterations_completed += 1
+        if isinstance(self._trigger_info.number, list):
+            self._frames_collected += self._trigger_info.number[
+                self._current_number_index
+            ]
+        else:
+            self._frames_collected += self._trigger_info.number
```

Comment:
Minor preference for the same comprehension as in `complete` 
```suggestion
        self._frames_collected += (
            self._trigger_info.number[self._current_number_index]
            if isinstance(self._trigger_info.number, list)
            else self._trigger_info.number
        )
```